src/vf2_layout.rs

```diff
@@ -0,0 +1,121 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use indexmap::IndexMap;
+
+use numpy::{PyReadonlyArray1, PyReadonlyArray2};
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+use rayon::prelude::*;
+
+use crate::nlayout::NLayout;
+
+const PARALLEL_THRESHOLD: usize = 50;
+
+/// Score a given circuit with a layout applied
+#[pyfunction]
+#[pyo3(
+    text_signature = "(bit_list, edge_list, error_matrix, layout, strict_direction, run_in_parallel, /)"
+)]
+pub fn score_layout(
+    bit_list: PyReadonlyArray1<i32>,
+    edge_list: IndexMap<[usize; 2], i32>,
+    error_matrix: PyReadonlyArray2<f64>,
+    layout: &NLayout,
+    strict_direction: bool,
```

Comment:
It's not necessarily reflected in the error matrix, it really depends on the backend properties since we could end up with an error rate defined directionally in the backend and it's hard to know until after we finish building the matrix so we don't build the error matrix bidirectional if `strict_direction=False`. It becomes more a question of scoring behavior than representing it. If `strict_direction = False` the and `err_mat[[0, 1]]` is `NaN` it will try `err_mat[[1, 0]]`

Reply:
Yeah, that's fair enough then.  In all the situations I could think of, the backend would just have constructed the directional error matrix itself, but if there's a chance those are decoupled, then it's right to include the swap.

That said, it feels a bit odd in the scoring that we don't take into account both sides of the link in other cases.  If it's not strict directionality, but the two directions have different average errors, it feels weird that we don't take that into account somehow?  In my mental model, that situation shouldn't be possible, but if the error matrix and strict directionality aren't coupled at the level of the backend, it starts to feel possible and unclear in how it should be handled.

Reply:
It's a tradeoff to find a path vs not finding a path. There are two modes of operation right now with the vf2 mapping, either we work with directed graphs and respect the direction of gates on the backend or we treat every edge as weak and work with undirected edges and then rely on `GateDirection` later to correct things. In the former case if the edge is defined bidirectionally with different error rates rustworkx will return a different mapping with either direction and score them differently. But for the latter case we can't really make an assertion about the order of the qubits for the qargs because it's not explicitly set.

That being said I think you're right for the undirected case maybe we should be looking at the other direction in scoring if both directions have defined error rates. We'll have to think of the best way to do this. (for this I just copied the scoring algorithm we were using before anyway)